Modularization of routes/logger/helpers

Mazdaratti opened the pull request:
- Separated HTML, API and Error-handler routes into dedicated blueprints for better modularity.
- Created logger module and imported into html and api routes
- Cleaned up project structure for better organization: moved helpers and routes into separate folders.

Mazdaratti committed:
Modularization of routes/logger/helpers

- Separated HTML, API and Error-handler routes into dedicated blueprints for better modularity.
- Created logger module and imported into html and api routes
- Cleaned up project structure for better organization: moved helpers and routes into separate folders.

Mazdaratti committed:
Update app.py

Mazdaratti committed:
Updated docstrings/removed unused imports

wagnerd commented on helpers/api_helpers.py:
```diff
@@ -0,0 +1,14 @@
+from flask import jsonify
+
+
+def create_response(status, message=None, data=None):
+    response = {"status": status}
+    if message:
+        response["message"] = message
+    if data:
+        response["data"] = data
+    return jsonify(response)
```
You can even go further and save multiple of those 'error'/'success' strings

```suggestion
def create_succes_response(message=None, data=None):
    return create_response('sucess', message, data)

def create_error_response(message=None, data=None):
    return create_response('error', message, data)


def create_response(status, message=None, data=None):
    response = {"status": status}
    if message:
        response["message"] = message
    if data:
        response["data"] = data
    return jsonify(response)
```

Mazdaratti replied:
Great suggestion! Thanks!

wagnerd commented on datamanager/sqlite_data_manager.py:
```diff
@@ -263,6 +262,8 @@ def update_movie(self, user_movie_id, updated_details):
         """
         try:
             user_movie = self.get_user_movie(user_movie_id)
+            if not user_movie:
+                return {"error": "This movie is not in your list."}
```
Just to show you there is the 'walrus operator' since Python 3.8 where you can write this even more compact if you want (https://docs.python.org/3/whatsnew/3.8.html):

```suggestion
            if not user_movie := self.get_user_movie(user_movie_id):
                return {"error": "This movie is not in your list."}
```

wagnerd commented on routes/api_routes.py:
```diff
@@ -0,0 +1,227 @@
+    """
+    api = Blueprint('api', __name__)
+
+    @api.route('/users', methods=['GET'])
+    def get_users():
+        """
+        Retrieves all users from the database.
+
+        Returns:
+            Response: JSON response containing a list of all users or an error message.
+        """
+        try:
+            users = data_manager.get_all_users()
+            return create_response("success", data=[user.to_dict() for user in users])
+        except Exception as e:
+            logger.error(f"Error in get_users: {str(e)}")
+            return create_response("error", message=str(e)), 500
```
Example comment for all the use cases:

If you have multiple times this same exception handling with the same messages and return codes, you can think about creating a decorator for this.

I didn't test it honestly but I think you get the idea :) 
This removes a lot of repeated error handling.

You can then think about having decorators for different exception types and error codes like 500 here etc.

```suggestion

from functools import wraps

def exc_handling(a_func):
    @wraps(a_func)
    def wrapTheFunction():
        try:
            a_func()
        except Exception as e:
            logger.error(f"Error in {a_func.__name__}: {str(e)}")
            return create_response("error", message=str(e)), 500
    return wrapTheFunction

    @exc_handling
    @api.route('/users', methods=['GET'])
    def get_users():
        """
        Retrieves all users from the database.

        Returns:
            Response: JSON response containing a list of all users or an error message.
        """
            users = data_manager.get_all_users()
            return create_response("success", data=[user.to_dict() for user in users])

```

Mazdaratti replied:
Very smart! I wish I come up on this idea by myself!

wagnerd reviewed (commented):
I had a rough look over the PR.
Never used flask more than just some experiments but left you some general ideas!

Overall looks really great the whole repository. Nice comments, docstrings and structure!
Always nice to look at your code :)

Mazdaratti committed:
Optimized helper for api responses

Mazdaratti committed:
Docstrings added

Mazdaratti committed:
Implemented walrus operator/small improvements

Mazdaratti committed:
Created decorators for api_routes

Mazdaratti committed:
Implemented decorators in routes

- decorator to handle exceptions and standardize API error responses
- decorator to validate JSON request data for API routes

Mazdaratti committed:
decorator.py ->api_decorator.py

Mazdaratti committed:
Error handling decorator for html routes created

Mazdaratti committed:
Decorator applied

Mazdaratti committed:
Project structure updated

Mazdaratti committed:
Update movieweb.sqlite
